### ISSUE TITLE
Status of ctags parsers in Geany

### ISSUE THREAD

**Issue by techee** [geany/geany#3170]
I more or less pushed all the remaining ctags parsers in the recent PRs so here's a short summary of the remaining diffs:

* geany_docbook.c: I think @masatake wanted to make this a proper XML parser in https://github.com/universal-ctags/ctags/pull/2198 which isn't probably interesting for us as we'd need to introduce a dependency on some XML parser. In any case, the parser is simple and there should be no problem to maintain it by ourselves.
* geany_matlab.c: Upstream uctags uses a regex-based parser which extracts some more tags, on the other hand regex parsers tend to be a bit slow so I'd stick with the current (again simple) parser for Geany (right now I don't feel very motivated to improve this parser so it can replace the regex-based parser).

geany_c.c (used as a parser for Java, C#, D, Vala) and geany_lcpp.c (preprocessor for geany_c.c):

1. I'd like to first simplify these by removing the C, C++, Vera parsers which we don't use so there's less code to compare. I already did this in uctags in https://github.com/universal-ctags/ctags/pull/3332 and I'd do something similar in Geany. 
2. Next, I would try to modify it so it uses `cpreprocessor.c` instead of the `geany_lcpp.c` preprocessor. The biggest difference is how function signatures are collected, I'd probably use the uctags implementation.
3. We should then decide what to do with the Vala parser which is present in Geany but not in uctags c.c. There's a PR in uctags https://github.com/universal-ctags/ctags/pull/2677 implementing this parser but I'm not sure how ready it is. Another alternative is to update universal ctags c.c with Vala. Yet another option is a hybrid approach where we could extract Vala from c.c in a similar way to (1), keep using it and eventually migrate to uctags Vala implementation when it is ready.
4. Finally, the rest will have to be synced piece by piece.

**Comment by elextr** [comment]
Docbook: the new ctags requirement for well formed XML will probably make symbols disappear while the input is illegal as the user types, so the symbols pane will flash, and symbols autocomplete won't be available.  So keeping current one may indeed be better.

Matlab: nearly every other tool uses Regex parsers successfully, I doubt matlab files are big, IIUC they tend to be lots of small ones.

`C.C`: [shudder] whats the state of ctags for Java, C#, D, Vala?  If they are likely to be similar to Geany in capability "soon" why remove C, C++, Vera and make the other changes? Why not just leave it there until we can finally drive a stake through `C.C`? [trying to save you work]

**Comment by techee** [comment]
> Matlab: nearly every other tool uses Regex parsers successfully, I doubt matlab files are big, IIUC they tend to be lots of small ones.

Well, I don't care much myself, we could switch to the regex parser too. (On the other hand I find it quite nice that we got rid of all the regex parsers so this one would be the only outlier.)

> C.C: [shudder] whats the state of ctags for Java, C#, D, Vala? If they are likely to be similar to Geany in capability "soon" why remove C, C++, Vera and make the other changes? Why not just leave it there until we can finally drive a stake through C.C? [trying to save you work]

So I can sync geany_c.c and ctags c.c more easily and don't see the the C, C++ and Vera garbage we don't care about.

**Comment by elextr** [comment]
Not sure this is what this issue is for, but it makes a convenient summary place.

I went for it and compiled ran tests which all passed (as expected with @techee PRs) and (minimally) tested symbols with random code in the language from github for all PRs with "ctags parser" label on them except:

1. #3034 Kotlin has conflicts
2. #3160 Fortran is waiting to be updated with changes from Geany's old parser that are being pushed upstream
3. #3169 Raku is a whole filetype, not just grabbing upstream parser, will look at it separately

The results are:

1. #3157 TCL procedure symbols show
2. #3158 VHDL symbols show
4. #3159 Latex symbols show
5. #3161 Markdown document outline shows

Run out of time, will do more later

**Comment by elextr** [comment]
Uh Oh, though tests pass, no symbols found for Clojure, even for `tests/ctags/simple.clj`

**Comment by techee** [comment]
I'm not sure whether we want to merge #3034. It really seemed to be quite slow on the Raspberry (but maybe it isn't such an issue, I don't know).

**Comment by techee** [comment]
> Uh Oh, though tests pass, no symbols found for Clojure, even for tests/ctags/simple.clj

Hmm, I can see them in the symbol tree. Are you sure you compiled the correct PR? (Also, if you somehow merged all the PR into a single branch, there will probably be conflicts that have to be resolved because each of the new parsers is added to the end of the parser list separately).

**Comment by elextr** [comment]
> (Also, if you somehow merged all the PR into a single branch, there will probably be conflicts that have to be resolved because each of the new parsers is added to the end of the parser list separately).

No each was in its own branch, and anyway there were no conflicts applying them, when they didn't work I actually went back and checked the pull and it was fine.  Are there any other PRs (not labelled ctags) needed for them to work?  The tests pass, but `simple.clj` shows no symbols, so maybe the tests are also faulty?

I built all previously with meson, so I made two new clones compiled one with autotools, one with meson, applied just #3162 to both and built and both work, so its not meson.  Maybe some interaction, git did clojure as a fast forward not a merge as the other ones were?  I just copy and paste step 1 of the "command line instructions" on the PR to the terminal, nothing special.

Also checked with a random selection from [here](https://github.com/clojure/clojure/tree/master/test/clojure/test_clojure) and seemed to work, so I would say #3162 is fine by itself, I guess we see how it goes when they are all merged together.

Having had to wait for an autotools build I am now out of time :stuck_out_tongue_winking_eye: , will test others tomorrow.

**Comment by techee** [comment]
@elextr Have you possibly overridden the default `filetypes.Clojure.conf` with your own configuration? Because this is not a built-in filetype, I just enabled the parser by adding `tag_parser=Clojure` under the `[Settings]` section (we do the same with the JSON parser for example) and your custom config file may miss it.

**Comment by elextr** [comment]
@techee, have pinged the PRs that I successfully tested so we can let C++ cool off for a bit.


On Clojure, no I didn't change the config, everything was done in a new clone with a new prefix and using -c for a new config, but as I said [above](https://github.com/geany/geany/issues/3170#issuecomment-1104557197) Clojure works when tested by itself, lets see how it goes on merge to master.

**Comment by techee** [comment]
> On Clojure, no I didn't change the config, everything was done in a new clone with a new prefix and using -c for a new config, but as I said https://github.com/geany/geany/issues/3170#issuecomment-1104557197 Clojure works when tested by itself, lets see how it goes on merge to master.

This is weird though and we should figure out what is going on in this case and despite tests passing, for users this is a non-working parser.

By the way, I haven't tried but how do the filetype config files behave? Is it so that the installed config file is loaded and then the user config file from `.config` is loaded overriding the corresponding keys of the installed config file? Or does Geany behave so that when a file from `.config` is found, the installed config file isn't read at all? I'm wondering because if Geany behaves the second way, even if we add `tag_parser=Clojure` to the config file, it wouldn't get enabled for users having their custom config file.

> In fact no symbols for any of the higher numbered ctags PRs???

For what other parsers do you see this problem?

**Comment by elextr** [comment]
> This is weird though and we should figure out what is going on in this case and despite tests passing, for users this is a non-working parser.

Agreed its weird, but as I said in the comment above when your branch was added by itself it worked, both tests and on screen symbols.  I applied the PRs in numerical order so. Maybe merge the PRs with lower numbers and then we can test the clojure PR again before merging it.  I left merge to you since you said above there was the possibility of conflicts in the parsers list when they are all merged to master.

**Comment by techee** [comment]
> Agreed its weird, but as I said in the comment above when your branch was added by itself it worked

Ah, OK, I thought it happened also separately. Then I think it's because the parsers were added to the end of the parser list, each parser to the same position and this could introduce some problems when you applied multiple patches.

> I left merge to you since you said above there was the possibility of conflicts in the parsers list when they are all merged to master.

Yeah, I think after merging the first parser, all the subsequent parsers will get merge conflicts.
